Feature/loadtesting v2023 04 01 preview/draft1

Himanshu49 opened the pull request:
# Data Plane API - Pull Request
<!-- 🚨🚨🚨 Replace this line with a summary and reason for these changes to your data plane API 🚨🚨🚨 -->

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [x] a public preview
- [x] GA release

### Change Scope

<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Added a new API version - 2023-04-01-preview
* Design Document:
* Previous API Spec Doc: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/loadtestservice/data-plane/Microsoft.LoadTestService/stable/2022-11-01/loadtestservice.json
* Updated paths: No root path updated.

### Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/85/Swagger-Suppression-Process) 
to get approval.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://aka.ms/style)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

fix https://github.com/Azure/azure-rest-api-specs/issues/26367

catalinaperalta committed:
initial tsp

catalinaperalta committed:
client.tsp

catalinaperalta committed:
client.tsp improvements

catalinaperalta committed:
model updates

catalinaperalta committed:
update routes

catalinaperalta committed:
update implementation

MaryGao committed:
Sync typespec and swagger change for loadtesting service (#24412)

* Update the diff for put and patch operation

* Change LoadTestAdministrationOperations to LoadTestAdministration to align with the swagger LoadTestAdministration_ prefix

Change LoadTestRunOperations to LoadTestRun to align with the swagger LoadTestRun_ prefix

catalinaperalta committed:
update models

catalinaperalta committed:
add list metric dimensions op and fix create test run

catalinaperalta committed:
update client.tsp

catalinaperalta committed:
fix DimensionValueList and update TestRun model

MaryGao committed:
Configuration changes for LoadTesting (#24481)

* Update the projected name

* Renamed

* update config file

* update folder

* Remove client.tsp

* Update the projected name to old ones

* Adjust the positions

* Update client.tsp

* Update tspconfig.yaml

Added correct options for python emitter.

* Update the metadata

* Change the projectedName to client.tsp

* Remove the projectedName in routes.tsp

* update file

* Update the .Net config

* Remove the projectedName change for load testing

* Revert position change

* Update specification/loadtestservice/LoadTestService/tspconfig.yaml

---------

Co-authored-by: Ray Chen <raychen@microsoft.com>

catalinaperalta committed:
update routes to be more accurate

catalinaperalta committed:
update client.tsp for python

catalinaperalta committed:
Merge branch 'feature/loadtesting' of https://github.com/Azure/azure-rest-api-specs into feature/loadtesting

catalinaperalta committed:
Merge branch 'main' of https://github.com/Azure/azure-rest-api-specs into feature/loadtesting

msyyc committed:
python config to avoid breaking (#24801)

* python config

* Update client.tsp

weidongxu-microsoft committed:
java, feature/loadtesting, match tsp and swagger (#24878)

* java config

* project for java

* metricNamespace is required

* aggregation is optional

* body is optional

* metricName and timespan is required

Himanshu49 committed:
Changes for 2023-04-01-preview version

Himanshu49 committed:
Minor refactoring

Himanshu49 committed:
Added URL based changes and versioning support (versioning not working)

Himanshu49 committed:
Versioning support changes

Himanshu49 committed:
Refactoring

Himanshu49 committed:
Refactoring

Himanshu49 committed:
config changes

Himanshu49 committed:
Added examples and few refactoring

Himanshu49 committed:
Adding emitters for pythin, ts and csharp

Himanshu49 committed:
Updated description for timespan field

Himanshu49 committed:
Update tspconfig as validation pipelines failing.

Himanshu49 committed:
Rename createOrUpdateTest.json to CreateOrUpdateTest.json

Himanshu49 committed:
Rename createOrUpdateTestRun.json to CreateOrUpdateTestRun.json

Himanshu49 committed:
Rename createOrUpdateTest.json to CreateOrUpdateTest.json

Himanshu49 committed:
Rename createOrUpdateTestRun.json to CreateOrUpdateTestRun.json

Himanshu49 committed:
suppress warning in client.tsp

Himanshu49 committed:
Merge branch 'feature/loadtesting-v2023-04-01-preview/draft1' of https://github.com/Himanshu49/azure-rest-api-specs into feature/loadtesting-v2023-04-01-preview/draft1

Himanshu49 committed:
Updated examples for stable version

Himanshu49 committed:
Updated examples based on latest changes

Himanshu49 committed:
Updated spec for 2023-04-01-preview

Himanshu49 committed:
Added description for APIVersions model

catalinaperalta commented on specification/loadtestservice/LoadTestService/models.tsp:
```diff
@@ -0,0 +1,1107 @@
+}
+
+@doc("Collection of tests")
+@friendlyName("TestsList")
```
I dont think you need `@friendlyName` here.

catalinaperalta replied:
Same comment in other places where `friendlyName` is being used.

Himanshu49 replied:
We added @friendlyName because the models were getting generated with the name similar to Paged{ResourceName} in this case PagedTest. In our existing spec we have name this model as TestsList. Should I remove this decorator and keep these models name to default?

catalinaperalta replied:
Ah I see, I think `@projectedName` would be better here. friendlyName is more for models that are templates, not ones that use templates. Here's the link to info on projected names: https://microsoft.github.io/typespec/standard-library/built-in-decorators#@projectedName

catalinaperalta replied:
I see this is still as friendlyName, are you planning on updating it to projectedName?

Himanshu49 replied:
@doc("The response to a metrics query.")
@projectedName("json", "Metrics")
//@friendlyName("Metrics")
model Metrics is Azure.Core.Page<TimeSeriesElement>;

tried this, but the json was still coming with PagedTimeSeriesElement. Although projectedName is working in case of parameter within models, as given in examples.

catalinaperalta replied:
@timotheeguerin just double checking before I resolve this comment, is friendlyName ok here?

catalinaperalta replied:
@Himanshu49 the issue is that the name isnt coming out as expected in the generated swagger correct? Have you tried generating an SDK with this to see what it outputs?

timotheeguerin replied:
This is an issue with `@friendlyname` we are discussing and why you shouldn't use `@friendlyname` in general but here yeah this is probably the only way to work around it right now.

catalinaperalta replied:
Thanks Tim! @Himanshu49 let's leave it with friendlyName for now if that's working for you

timotheeguerin replied:
This is the issue if you want to link it https://github.com/microsoft/typespec/issues/2276

catalinaperalta commented on specification/loadtestservice/LoadTestService/models.tsp:
```diff
@@ -0,0 +1,1107 @@
+
+// @doc("Load test run model")
+// @resource("test-runs")
+// model TestRuns {
```
Remove this model

catalinaperalta commented on specification/loadtestservice/LoadTestService/routes.tsp:
```diff
@@ -0,0 +1,692 @@
+  @route("/tests")
+  @get
+  @example("./examples/ListTests.json", "ListTests")
+  listTests is Azure.Core.Foundations.Operation<
```
This operation didn't fit into a ResourceList operation?

Himanshu49 replied:
Updated as suggested

catalinaperalta commented on specification/loadtestservice/LoadTestService/routes.tsp:
```diff
@@ -0,0 +1,692 @@
+  @route("/test-runs/{testRunId}")
+  @get
+  @example("./examples/GetTestRun.json", "GetTestRun")
+  getTestRun is Azure.Core.Foundations.Operation<
```
Can the get operations be defined as ResourceRead operations?

Himanshu49 replied:
Updated as suggested.

catalinaperalta reviewed (commented):
A few high-level comments: 
- Make sure to review this doc about how to structure TypeSpec projects that are checked into this repo: https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/typespec-structure-guidelines.md
- Following the structure recommendations, the openapis should be checked in under a `data-plane` folder. 
- Also the examples should be at the same level as your tsp files. 
- Remove tsp-output/
- Can more operations use the recommended TypeSpec Azure templates?

catalinaperalta commented:
Also make sure to check the TypeSpec Validation step and address any issues there after you've pushed more recent commits.

Himanshu49 commented on specification/loadtestservice/LoadTestService/models.tsp:
```diff
@@ -0,0 +1,1107 @@
+  duration?: int32;     
+}
+
+// @doc("Optional load test config")
```
Yes, we were in discussion with API review team on how to move forward on our implementation causing a breaking change. Will remove this and other commented code.

Himanshu49 committed:
Updated readme and added examples

Himanshu49 committed:
removed tsp-output folder and files in it

Himanshu49 committed:
Updated service name in client

Himanshu49 committed:
Updated api version parameter

Himanshu49 committed:
Pushing updated json

Himanshu49 committed:
Removing unused example file

Himanshu49 committed:
suppressing warnings in client.tsp

Himanshu49 committed:
Changes to use standard templates for operations

Himanshu49 committed:
Changing create and update to generic template as it is changing to default resource parameter.

Himanshu49 committed:
Updating visibility of FileName

Himanshu49 committed:
Minor change in name

Himanshu49 committed:
changes in visibility of filename

Himanshu49 committed:
Change in model name of fileInfo

Himanshu49 committed:
Changing file info to testfileinfo

Himanshu49 committed:
Updated jsons

Himanshu49 committed:
Changes to rectify warning in pipeline

Himanshu49 committed:
changing metricName to metricname

Himanshu49 committed:
Updated examples to revert to metricname

Himanshu49 committed:
Ran tsp format and update example for metricname parameter

Himanshu49 committed:
Added query parameter for list test and testrun operation

Himanshu49 committed:
Updated description

prativen commented on specification/loadtestservice/LoadTestService/main.tsp:
```diff
@@ -0,0 +1,48 @@
+  }
+)
+@doc("""
+These APIs allow end users to create, view and run load tests using Azure Load
```
nit: can it be in same line

prativen commented on specification/loadtestservice/LoadTestService/models.tsp:
```diff
@@ -0,0 +1,1114 @@
+@doc("Metrics for pass/fail criteria.")
+enum PFMetrics {
+  @doc("Pass fail criteria for response time metric.") response_time_ms,
+  @doc("Pass fail criteria for response time metric.") latency,
```
incorrect comment. Latency metric. Also what is the unit

prativen commented on specification/loadtestservice/LoadTestService/models.tsp:
```diff
@@ -0,0 +1,1114 @@
+@doc("Types of secrets supported.")
+enum SecretType {
+  @doc("If the secret is stored in an Azure Key Vault.") AKV_SECRET_URI,
+  @doc("If the Plain text secret value provided.") SECRET_VALUE,
```
nit: plain

mikeharder committed:
Merge branch 'main' into feature/loadtesting-v2023-04-01-preview/draft1

mikeharder committed:
Enable linter rules

Himanshu49 committed:
Updated description of models and operations

Himanshu49 committed:
Updated specs after internal review of APIs

Himanshu49 committed:
Added ContainerInfo in 2023-04-01-preview

Himanshu49 committed:
Updated operation id for stop test run

Himanshu49 committed:
Updated operation id for stop test run

Himanshu49 committed:
Updated default values of properties

Himanshu49 commented:
/azp run

Himanshu49 committed:
Description changes for few properties and models

Himanshu49 committed:
Addressed errors in breaking change pipeline

Himanshu49 committed:
Merge branch 'main' into feature/loadtesting-v2023-04-01-preview/draft1

mikekistler commented on specification/loadtestservice/LoadTestService/routes.tsp:
```diff
@@ -0,0 +1,488 @@
+
+@tag("Test")
+interface LoadTestAdministration {
+  #suppress "@azure-tools/typespec-azure-core/use-standard-operations"
+  @summary("Create a new test or update an existing test by providing the test Id.")
+  @doc("Create a new test or update an existing test by providing the test Id.")
+  @route("/tests/{testId}")
+  @patch
+  createOrUpdateTest is Azure.Core.Foundations.Operation<
```
Why couldn't you use `ResourceOperations.ResourceCreateOrUpdate` here?

Himanshu49 replied:
Because it is changing the request body. We expect a json body in a parameter named body. After using standard operation, it changes to parameter name "resource".

mikekistler replied:
If it only changes the parameter name then it is still preferable to use the Azure Core operation template.

Himanshu49 replied:
We have sdk based out of these specs which expects named parameter body in operations. Have raised an issue - https://github.com/Azure/typespec-azure/issues/3742 regarding this.

catalinaperalta replied:
Given the discussion in the issue (https://github.com/Azure/typespec-azure/issues/3742) let's leave this as is for now and update later on if the template supports updating the parameter name

mikekistler commented on specification/loadtestservice/LoadTestService/routes.tsp:
```diff
@@ -0,0 +1,488 @@
+""")
+  @route("/tests/{testId}/files/{fileName}")
+  @put
+  uploadTestFile is Azure.Core.Foundations.Operation<
```
Why can't this be `ResourceOperations.ResourceCreateOrReplace`?

Himanshu49 replied:
This is because creating a file requires content of file as body and response is test file information (json). So, since the input and output is customized in case of put, therefore using generic template for this one as well.

Himanshu49 replied:
Following it with this issue - https://github.com/Azure/typespec-azure/issues/3743

catalinaperalta commented on specification/loadtestservice/LoadTestService/client.tsp:
```diff
@@ -0,0 +1,63 @@
+  service: Microsoft.LoadTestService,
+})
+interface AdministrationOperations {
+  CreateOrUpdateTest is LoadTestAdministration.createOrUpdateTest;
```
Same comment other operations/models where this applies

Himanshu49 replied:
Our already published GA sdk have these operation names in Pascal case, would changing the operation name in client.tsp to camel case create a breaking change for sdk versions?

mikekistler replied:
It should not. The generators all munge the operation names into the idiomatic casing for the language, so you should be safe to use camel casing in the TypeSpec.

catalinaperalta commented on specification/loadtestservice/LoadTestService/client.tsp:
```diff
@@ -0,0 +1,63 @@
+  service: Microsoft.LoadTestService,
+})
+interface TestRunOperations {
+  // TODO check with service team why is Python different here?
```
Let's remove these comments if you have verified that everything is correct

catalinaperalta commented on specification/loadtestservice/LoadTestService/main.tsp:
```diff
@@ -0,0 +1,47 @@
+  title: "Azure Load Testing",
+})
+@server(
+  "https://{endpoint}",
```
The endpoint folks get from the portal doesnt include the scheme?
```suggestion
  "{endpoint}",
```

Himanshu49 replied:
Our resource endpoint from portal doesn't include scheme, therefore added https://

catalinaperalta commented on specification/loadtestservice/LoadTestService/models.tsp:
```diff
@@ -0,0 +1,1049 @@
+  name?: string;
+
+  @doc("The namespace the metric belongs to.")
+  `namespace`?: string;
```
Why is this between quotes?

Himanshu49 replied:
because namespace is a keyword in TypeSpec.

catalinaperalta commented on specification/loadtestservice/LoadTestService/tspconfig.yaml:
```diff
@@ -0,0 +1,50 @@
+    package-dir: "azure-developer-loadtesting"
+    package-name: "{package-dir}"
+    # emitter-output-dir: "{python-sdk-folder}/sdk/{service-directory-name}/{package-name}"
+    package-mode: dataplane
```
I think the latest emitter version is now expecting `azure-dataplane`

mikekistler commented on specification/loadtestservice/LoadTestService/tspconfig.yaml:
```diff
@@ -0,0 +1,50 @@
+    emitter-output-dir: "{project-root}/../"
+    output-file: "{azure-resource-provider-folder}/{service-name}/{version-status}/{version}/loadtestservice.json"
+    azure-resource-provider-folder: "data-plane"
+    examples-directory: ./examples
```
Recommend you configure autorest to omit unreachable types:
```suggestion
    examples-directory: ./examples
    omit-unreachable-types: true
```

mikekistler commented on specification/loadtestservice/LoadTestService/models.tsp:
```diff
@@ -0,0 +1,1049 @@
+@doc("Test file info.")
+@resource("files")
+@parentResource(Test)
+model TestFileInfo extends FileInfo {
```
I think `is` would be better than `extends` here. You don't really want inheritance. Might also move the fileName into FileIno so it does not have to be declared twice.
```suggestion
model TestFileInfo is FileInfo {
```

Himanshu49 replied:
Thanks for the suggestion, updated the code.

mikekistler commented on specification/loadtestservice/LoadTestService/models.tsp:
```diff
@@ -0,0 +1,1049 @@
+  resourceType: string;
+
+  @doc("Azure resource display name")
+  displayName: string;
```
Is `displayName` required? I see that the descriptions for `resourceName` and `resourceType` say they are required, but no for `displayName`.

Himanshu49 replied:
Display name is not required for this model. Removing it from required list.

mikekistler commented on specification/loadtestservice/LoadTestService/models.tsp:
```diff
@@ -0,0 +1,1049 @@
+
+  @doc("Test run duration in milliseconds.")
+  @visibility("read")
+  duration?: int32;
```
In the GA API this is declared as `int64`. Is there some reason we need to change it here?

Himanshu49 replied:
Missed it Mike, changed it back to int64.

mikekistler reviewed (changes requested):
Please address outstanding comments and then tag us to re-review.

Himanshu49 committed:
Removed unused models, refactored code, addressed PR comments

Himanshu49 commented:
@mikekistler @catalinaperalta , We have addressed the comments and resolved most of them. Have dropped few comments on unresolved comments. Please re-review the latest code.

Commit:
nit: changes in client and model tsp

Commit:
spec after nit changes

Commit:
Changes suggested in API review

timotheeguerin commented on specification/loadtestservice/LoadTestService/client.tsp:
```diff
@@ -24,7 +24,7 @@ interface AdministrationOperations {
   listTestFiles is LoadTestAdministration.listTestFiles;
   listTests is LoadTestAdministration.listTests;
-  #suppress "@azure-tools/typespec-azure-core/byos"
+  #suppress "@azure-tools/typespec-azure-core/byos" //It is limiting for users to have storage first before creating a load test also it will be breaking change for us.
```
```suggestion
  #suppress "@azure-tools/typespec-azure-core/byos" "It is limiting for users to have storage first before creating a load test also it will be breaking change for us."
```

suppress takes a 2nd argument which is the message no need to use comment

Himanshu49 committed:
Update specification/loadtestservice/LoadTestService/client.tsp

Co-authored-by: Timothee Guerin <timothee.guerin@outlook.com>

Commit:
Suggested changes in API review

Commit:
Merge branch 'feature/loadtesting-v2023-04-01-preview/draft1' of https://github.com/Himanshu49/azure-rest-api-specs into feature/loadtesting-v2023-04-01-preview/draft1

Himanshu49 committed:
updating name and description for ContainerInfo

Himanshu49 committed:
Added publicIPDisabled property to Test and TestRun model in 2023-04-01-preview version

Himanshu49 committed:
Update properties in examples JSON files

Himanshu49 committed:
Linting changes

mikekistler reviewed (approved):
Looks good for preview. 👍

haolingdong-msft commented on specification/loadtestservice/LoadTestService/routes.tsp:
```diff
@@ -0,0 +1,438 @@
+import "@typespec/http";
+import "@typespec/openapi";
+import "./models.tsp";
+import "@azure-tools/typespec-autorest";
```
Why import "@azure-tools/typespec-autorest"? I think this is for your to config in `tspconfig.yaml`'s `emit` to generate swagger. Why will code use this lib?

Himanshu49 replied:
This is used because without this we were not able to use @example decorator to attach an example to one of our operation, as it has a custom OperationId. Please refer to stop operation in line 236 of routes.tsp file.

timotheeguerin replied:
Why do you have a custom operation id though? Either you don’t want to rename it from the previous swagger and then you should have the operation name be what is matching. Or you do want to rename then why have operation id?

timotheeguerin replied:
Also it should still resolve the examples automatically.

Himanshu49 replied:
Yes, we have an already published GA version which we are onboarding to TypeSpec, which already has SDK version released for them. We don't want to rename it. It is a custom action stop but its operation ID is LoadTestRun_StopTestRun instead of LoadTestRun_Stop which is generated by the typespec.

Himanshu49 replied:
Meanwhile I agree that example tag is not needed even if we use operation tag . Removed its reference and related import statements.

Himanshu49 committed:
Removed example tag and related imports

Himanshu49 committed:
Minor linting

Himanshu49 committed:
Merge branch 'main' into feature/loadtesting-v2023-04-01-preview/draft1

Himanshu49 committed:
Merge main and changes after run npx command in local

tjprescott commented:
@Himanshu49 is this PR ready to merge?

Himanshu49 commented:
> @Himanshu49 is this PR ready to merge?

Yes @tjprescott , we got the spec reviewed and are good to go.
